tiledb/api/src/array/dimension/mod.rs

```diff
@@ -223,6 +222,35 @@ impl<'ctx> Builder<'ctx> {
         })
     }
 
+    pub fn new_string(
```

Comment:
Ah, the one from the discussion yesterday.
A dimension has a domain and extent if and only if it is a dimension of integral or floating point type for a dense array... right?

And there's no indication in the API here that it's a dense array. Maybe there should be, maybe not - definitely would be a bit of a pain to re-do a bunch of the types to enforce these sorts of constraints. Might improve the code quality of the proptests.

Anyway I'm definitely not suggesting you do that, but if I got the "if and only if" right, then `new_string` is not a good name for this.  If it's optional then we should probably just pull it out of `Builder::new`, except we're beholden to calling `tiledb_dimension_alloc` with the domain/extent already known, so probably that means just using an Option arg on `new`?  `Option<DomainAndExtent>` or something, which has variants for each type so that there's no generic arg for the `new` function.

Reply:
This possible implementation of the `DomainAndExtent` is maybe a little annoying but not too pathological:
```
pub enum DimensionType {
    Int8(domain, extent),
    Int16(domain, extent),
    ...
    StringAscii
}

impl DimensionDatatype {
    pub fn datatype(&self) -> Datatype;
}

impl TryFrom<Datatype> for DimensionDatatype {
    ...
}
```
If we ever add variants to `Datatype` then they would be caught by a `match` clause in one of these functions, one assumes

Reply:
The if-and-only-if applies to the domain. There are also cases where extent is optional on non-string types, if memory serves I believe that's also sparse arrays so they can support floating point dimensions? Or perhaps any sparse dimension doesn't require an extent? Not sure on the particulars there. But easy enough to use your pattern there and just specify `Option<extent>`

Reply:
Actually, are you ok if I file a follow up story for this bit? I went to start implementing this and realized that it'll spider pretty far if I start tweaking the DimensionData class to be strongly typed. I absolutely think we should do that, but I'm thinking it s a bit much for this PR.

Reply:
To be clear, I mean specifically changing the DimensionData domain and extent fields from serde_json values into strongly typed Range values. (Copying from my mixup. Leaving the original to avoid confusion which this probably only adding to).

Reply:
Follow-up story sounds good, it's tricky to get things right with all the corner cases. Oof.

Reply:
Filed as [sc-46692]

Reply:
Looks like you can delete `new_string` and just use `new_raw`?

Reply:
From slack discussion -

The current difference is that `new_string` is `pub` so we need both to export the functionality to the example.  And the linked story is the plan to straighten this out so that further than that.